src/devTools.js

```diff
@@ -50,13 +60,41 @@ function connectToServer(options) {
   watch();
 }
 
-export function start(options) {
+
+// obtains hostname from the promise and then performs the callback
+// also, overwrites `optionsObject.hostname`
+// obtained hostname is stored in `obtainedHostname` variable
+// if it's already obtained, skip fetching
+async function obtainHostname(hostnamePromise, optionsObject, callback) {
```

Comment:
The point of making `obtainHostname` an async function was (1) so that we could avoid passing it any callback, (2) so the parent could handle `await`ing the `hostnamePromise`, and just pass the `hostname` in directly.

Would be worth spending some time reading up on async functions to get a better intuition about how they're used and how they relate to the concepts you're more familiar with (eg. promises and callbacks).

In this case, to get the full benefits here you should make the callers (`start`, `connect) `async function`s, have them handle `await`ing `hostnamePromise` and pass the `hostname` in here instead of `hostnamePromise`, and instead of a `callback` just have them call whatever they want after `obtainHostname` returns.

It's also not clear to me that we need to store `obtainedHostname` anywhere. As long as you get the same `hostnamePromise` passed in to `start` when it re-runs, it should immediately resolve with the same `hostname`. In other words, I think the `Promise` itself (passed in from our application layer) can store the state of the `hostname`.

Reply:
> The point of making obtainHostname an async function was (1) so that we could avoid passing it any callback, (2) so the parent could handle awaiting the hostnamePromise, and just pass the hostname in directly.

Not really. The reason why `obtainHostname` is `async` is that it is able to `await` for the passed promise's resolve. I introduced a callback here because the result of this promise is needed in two distinct places:
- `start` - this is invoked in `connect` and `send` functions, `send` is called from the object returned by `connect`, so, basically, all the calls are in the `connect` function.
- `connect` - this is problematic, because it's called in `connectViaExtension` which expects synchronously passed result from `connect`, and also it is being exported from the module and we use it in our code. Therefore I don't think making `connect` as `async` function is a good idea - it is against the whole concept of resolving the promise when we perform socket operations that are async anyway without making `async` all the functions on our way down the stack trace - if you make `connect async` you'll have to make `connectViaExtension` async as well as any function where you'd use `connect`'s result synchronously. In our case that'd be the code of `@react-navigation/devtools` [here](https://github.com/react-navigation/react-navigation/blob/ef7370b21585193bfc02ef03eddbf43f8c291151/packages/devtools/src/useReduxDevToolsExtension.tsx#L33). So in the end we'd have to patch `@react-navigation/devtools` again, right?

> It's also not clear to me that we need to store obtainedHostname anywhere. As long as you get the same hostnamePromise passed in to start when it re-runs, it should immediately resolve with the same hostname. In other words, I think the Promise itself (passed in from our application layer) can store the state of the hostname.

Yes, about that I do agree - this was introduced to handle the case when the promise's resolved to not "resolve" it anymore but I think you're right that the promise in such a case stores a resolved value and returns immediately so it is redundant.

I'm not 100% convinced by your solution but at the same time, I realized that this solution is a bit hacky and maybe it could be done better. I think making the `start` function `async` could be a way to go. I'm going to push a change with that. I realize that I might still be missing something so I'll just wait for your opinion about that.